server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java

```diff
@@ -112,69 +112,112 @@ public KillUnusedSegments(
+    stats.add(Stats.Kill.MAX_SLOTS, taskStats.maxSlots);
+  }
+
+  private int killUnusedSegments(
+      Collection<String> dataSourcesToKill,
+      int availableKillTaskSlots
+  )
   {
     int submittedTasks = 0;
-    for (String dataSource : dataSourcesToKill) {
-      if (submittedTasks >= availableKillTaskSlots) {
-        log.info(StringUtils.format(
-            "Submitted [%d] kill tasks and reached kill task slot limit [%d]. Will resume "
-            + "on the next coordinator cycle.", submittedTasks, availableKillTaskSlots));
-        break;
-      }
-      final Interval intervalToKill = findIntervalForKill(dataSource);
-      if (intervalToKill == null) {
-        continue;
-      }
-
-      try {
-        FutureUtils.getUnchecked(overlordClient.runKillTask(
-            TASK_ID_PREFIX,
-            dataSource,
-            intervalToKill,
-            maxSegmentsToKill
-        ), true);
-        ++submittedTasks;
-      }
-      catch (Exception ex) {
-        log.error(ex, "Failed to submit kill task for dataSource [%s]", dataSource);
-        if (Thread.currentThread().isInterrupted()) {
-          log.warn("skipping kill task scheduling because thread is interrupted.");
+    if (0 < availableKillTaskSlots && !CollectionUtils.isNullOrEmpty(dataSourcesToKill)) {
+      log.info("datasourcesToKill: %s", dataSourcesToKill);
+      for (String dataSource : dataSourcesToKill) {
+        if (submittedTasks >= availableKillTaskSlots) {
+          log.info(StringUtils.format(
+              "Submitted [%d] kill tasks and reached kill task slot limit [%d]. Will resume "
+              + "on the next coordinator cycle.", submittedTasks, availableKillTaskSlots));
           break;
         }
+        final Interval intervalToKill = findIntervalForKill(dataSource);
+        if (intervalToKill == null) {
+          continue;
+        }
+
+        try {
+          FutureUtils.getUnchecked(overlordClient.runKillTask(
+              TASK_ID_PREFIX,
+              dataSource,
+              intervalToKill,
+              maxSegmentsToKill
+          ), true);
+          ++submittedTasks;
+        }
+        catch (Exception ex) {
+          log.error(ex, "Failed to submit kill task for dataSource [%s]", dataSource);
+          if (Thread.currentThread().isInterrupted()) {
+            log.warn("skipping kill task scheduling because thread is interrupted.");
+            break;
+          }
+        }
       }
     }
 
-    log.debug("Submitted [%d] kill tasks for [%d] datasources.", submittedTasks, dataSourcesToKill.size());
+    log.debug(
+        "Submitted [%d] kill tasks for [%d] datasources.%s",
+        submittedTasks,
+        dataSourcesToKill.size(),
+        availableKillTaskSlots < dataSourcesToKill.size()
+            ? StringUtils.format(
+            " Datasources skipped: %s",
+            ImmutableList.copyOf(dataSourcesToKill).subList(submittedTasks, dataSourcesToKill.size())
```

Comment:
fwiw, I think this might need to be wrapped in an `log.isDebug()`, since the string format will likely be executed even if debug is not enabled.

Reply:
fixed